pyiron_contrib/tinybase/creator.py

```diff
@@ -0,0 +1,255 @@
+                    delete_aborted_job and job.status == "aborted"
+                ):
+                    job.remove()
+                else:
+                    if not isinstance(job.task, task):
+                        raise ValueError(
+                            f"Job with given name already exists, but is of different type!"
+                        )
+                    return job
+            except JobNotFoundError:
+                pass
+            return pyiron_contrib.tinybase.job.TinyJob(task(), self._project, name)
+
+        return create
+
+    def register(self, task: Union["AbstractTask", str], name: str):
+        """
+        Register a new task.
+
+        Args:
+            task (AbstractTask): new task to register
+
+        Returns:
+            AbstractTask: the newly registered task
+        """
+        if name is None:
+            name = type(task).__name__
+        if name in self._jobs:
+            if task != self._jobs[name]:
+                raise ValueError(
+                    "Refusing to register task: different task of the same name already exists!"
+                )
+            else:
+                return self._jobs[name]
+        self._jobs[name] = task
+        return task
+
+
+class TaskCreator(Creator):
+    def __init__(self, project, config):
+        self._tasks = config.copy()
+
+    def __dir__(self):
+        return tuple(self._tasks.keys())
+
+    def __getattr__(self, name):
+        task = self._tasks[name]
+        if isinstance(task, str):
+            task = self._tasks[name] = load_class(task)
+        return task
+
+    def register(self, task: Union["AbstractTask", str], name: str):
+        """
+        Register a new task.
+
+        Args:
+            task (AbstractTask): new task to register
+
+        Returns:
+            AbstractTask: the newly registered task
+        """
+        if name is None:
+            name = type(task).__name__
+        if name in self._jobs:
+            if task != self._jobs[name]:
+                raise ValueError(
+                    "Refusing to register task: different task of the same name already exists!"
+                )
+            else:
+                return self._jobs[name]
+        self._jobs[name] = task
+        return task
+
+
+class StructureCreator(Creator):
+    def __init__(self, project, config):
+        pass
+
+    @wraps(ase.build.bulk)
+    def bulk(self, *args, **kwargs):
+        return ase_to_pyiron(ase.build.bulk(*args, **kwargs))
+
+    @wraps(Atoms)
+    def atoms(self, *args, **kwargs):
+        return Atoms(*args, **kwargs)
+
+
+class ExecutorCreator(Creator):
+    _DEFAULT_CPUS = min(int(0.5 * len(sched_getaffinity(0))), 8)
+
+    def __init__(self, project, config):
+        self._most_recent = None
+
+    def most_recent(self):
+        if self._most_recent is None:
+            self._most_recent = Executor()
+        return self._most_recent
+
+    def _save(func):
```

Comment:
This gets away without a `staticmethod` decorator??

Reply:
Ah, right, because the `@_save` invocation also doesn't know about `self` or anything, it's just looking in the scope of the class. Cool. The only ugly part about it is that it's really line-order-dependent, so you can't just slide `_save` down below, it _has_ to appear before its invocations. That's not a high price or anything though.

Reply:
Yeah, `_save` is really just a convenience decorator, it could just as well be defined outside the class, probably that's cleaner anyway.